### ISSUE TITLE
oauth2_consent_scope_map not found in list of valid attributes

### ISSUE THREAD

**Issue by Flakebi** [kanidm/kanidm#2596]
### I did this
Upgrade from 1.1.0-rc.15 to 1.1.0-rc.16

### I expected the following
It works as flawlessly as all past upgrades

### Kanidm version details

- Output of `kanidm(d) version`: kanidm 1.1.0-rc.16
- Are you running it in a container? If so, which image/tag?: No
- If not a container, how'd you install it: I use the NixOS package and service
- Operating System / Version (On Unix please post the output of `uname -a`): `Linux <host> 6.4.0 #1-NixOS SMP PREEMPT_DYNAMIC Sun Jun 25 23:29:58 UTC 2023 x86_64 GNU/Linux`

### Any other comments
Log output:
```
migrate_domain_4_to_5 [ 5.51ms | 0.69% ]
┝━ ｉ [info]: Removing expired auth session | session_id: <uuid>
┝━ ｉ [info]: Removing expired oauth2 session | o2_session_id: <uuid>
┝━ ｉ [info]: Removing expired auth session | session_id: <uuid>
┝━ ｉ [info]: Removing expired oauth2 session | o2_session_id: <uuid>
┝━ ｉ [info]: Removing expired auth session | session_id: <uuid>
┝━ ｉ [info]: Removing expired oauth2 session | o2_session_id: <uuid>
┝━ 🚨 [error]: oauth2_consent_scope_map <user>@<domain> - not found in the list of valid attributes for this set of classes ["account", "memberof", "object", "posixaccount", "service_account"] - valid attributes are ["account_expire", "account_valid_from", "api_token_session", "class", "description", "directmemberof", "displayname", "entry_managed_by", "gidnumber", "jws_es256_private_key", "last_modified_cid", "loginshell", "mail", "memberof", "name", "name_history", "oauth2_session", "primary_credential", "spn", "ssh_publickey", "unix_password", "user_auth_token_session", "uuid"] | event_tag_id: 1
┕━ 🚨 [error]: Schema Violation in validation of modify_pre_apply AttributeNotValidForClass("oauth2_consent_scope_map") | event_tag_id: 1
```

As far as I understand the source, `oauth2_consent_scope_map` is part of the `account` class, so I don’t quite understand why it fails.

Downgraded to 1.1.0-rc.15 for now, which starts successfully.

**Comment by yaleman** [comment]
Have you possibly got an account and an oauth2 RS with the same name?

**Comment by tumbl3w33d** [comment]
I just wanted to open a similar ticket because I also ran into this issue and I already talked about it with @Firstyear and can contribute some more insights.

What happens is that the migration fails due to an unexpected/undesired attribute `oauth2_consent_scope_map`. In my case this happens with all service accounts that I used for interactively logging into applications (like Jenkins, gitea, …) in order to create an API token for programmatic access.

Apparently this use case wasn't obvious enough, so this ticket here brings it up. However, there is possibly already an alternative by the (recently, I think) implemented client credentials grant. I will explore that today on my end to figure out if it remove the necessity for the workflow I described before.

So to summarize:

- migration rc.15 -> rc.16 fails when (probably only service) accounts have such oauth2_consent_scope_map
  - this needs a fix either way
- if it turns out this (or similar) use cases like mine and possibly @Flakebi 's is valid, it might be an acceptable option to bypass the consent phase for service accounts

If there's an easy workaround to get rid of these consents with raw magic, that would take some pressure of this issue in my view because I could simply update without them and go the new way with client credentials grant (which is yet to be verified).

**Comment by yaleman** [comment]
You could try purging the attribute using raw (obviously use/take a backup first!)

```shell
echo '[{ "purged": "oauth2_consent_scope_map" }]' > change.json
kanidm raw modify '{"eq":["name", "yourtargetuser"]}' change.json -D admin@idp.example.com
```

**Comment by tumbl3w33d** [comment]
`admin` is not allowed
```
1cf532e9-7203-4cef-86f2-4e213f7db12a INFO           ┝━ ｉ [info]: requested_rem: {"oauth2_consent_scope_map"} !⊆ allowed: {"class", "displayname", "legalname", "mail", "name"} | event_tag_id: 11
1cf532e9-7203-4cef-86f2-4e213f7db12a INFO           ┕━ ｉ [info]: denied ❌ - modify may not proceed | event_tag_id: 11
```
and `idm_admin` neither
```
bae52e6d-c470-4655-a1a5-dafd88b4e634 INFO           ┝━ ｉ [info]: requested_rem: {"oauth2_consent_scope_map"} !⊆ allowed: {"account_expire", "account_valid_from", "api_token_session", "devicekeys", "displayname", "gidnumber", "id_verification_eckey", "legalname", "loginshell", "mail", "name", "passkeys", "primary_credential", "radius_secret", "ssh_publickey", "unix_password", "user_auth_token_session"} | event_tag_id: 11
bae52e6d-c470-4655-a1a5-dafd88b4e634 INFO           ┕━ ｉ [info]: denied ❌ - modify may not proceed | event_tag_id: 11
```

**Comment by yaleman** [comment]
CC grants are basically "[basic auth with POST](https://aaronparecki.com/oauth-2-simplified/#client-credentials)" which gives you a bearer token to use, because implicit was giving it back in the URL, through redirects, and that's terrible.

> The client credential grant seems to assume I have a client_id/client_secret pair, but it's not clear to me where to take it from. It's probably not the right idea that I recycle the pair that my OIDC-connected app uses in general for interactive logins. 

That's part of the changes that were made, that an OAuth2 RS is considered "an account", for use in this way - I'd say create another one for the use case.

> How would it be associated with the right application then? I mean, would my OIDC-application even have a use for the bearer token that I bring which is associated with a different oauth2 config in kani?

I'm lost. Does it support OIDC client credentials or not?

**Comment by tumbl3w33d** [comment]
No, there's no explicit support for it and I guess I was hoping it's something I could retrofit on these apps connected via auth code grant.

**Comment by Flakebi** [comment]
> Have you possibly got an account and an oauth2 RS with the same name?

Not sure what an oauth2 RS is, but if I search for my username in the json created by kanidm backup, I only find fields associated with the user and nothing else.

The `kanidm raw modify` command gives an access denied for me as well.

In case it matters, the account where the upgrade chokes is not a service account (it’s actually my account). I used it to login via oauth before.

**Comment by yaleman** [comment]
Sounds like a @Firstyear case! I tried digging into the code and couldn't work out where that'd be failing - we might need logs in debug mode to figure this one out, to see if there's something else going on.

**Comment by Firstyear** [comment]
@tumbl3w33d Had this same issue, I will investigate soon.

**Comment by Firstyear** [comment]
@Flakebi I'm really sorry your upgrade process wasn't as smooth as normal. I'm also sorry it took so long to respond, the last few days have been a bit rough for me. 

The issue is pretty simple (William is silly). I made an assumption which was incorrect that service accounts would not be using oauth2 tokens since how can a machine provide consent? I overlooked that some people like @tumbl3w33d and yourself may actually be using this flow.

As a result, you have service accounts with consent scopes applied because you have likely automated service account access to an oauth2 system. 

Now this raises a few options about how to proceed, and it depends on the use cases you have. I would like both of you (@Flakebi and @tumbl3w33d to describe your environments where you may be using a service account to access oauth systems so that I can refine my ideas better).

My current thoughts are: 

1) Allow direct issuance of oauth2 api tokens to a service account

This would solve @tumbl3w33d's problem, because they are currently using a service account with a password or api-token to then do an authorisation code grant to get an oauth2 token. There isn't really a point to this dance - we potentially could just issue the oauth2 token directly, allowing someone to use that with other resource servers. 

Where this idea has issues is that oauth2 makes a lot of assumptions about clients (those who hold the bearer token) and the resource servers (who receive a token and then verify it). By issuing a service account token directly, it may bypass some of this oauth2 logic around clients, and allow someone to directly access an rs. 

2) We allow service-accounts to bypass the scope consent 

This allows any existing flow you have to continue to work, but just skips the consent phase since a service-account doesn't need it.

3) Revert the change

We revert the change and allow oauth2 consent maps to remain on service accounts, and the requirements that an authorisation code flow must accept them. 


-- 

Regardless of the choice we make, longer term the better option here will be https://github.com/kanidm/kanidm/issues/2545 - this will allow multiple oauth2 clients to exist in a shared key domain. Effectively you would create one client for say "nextcloud" and another as your "service account". Because they have the same signing keys, a client credentials grant to the service account would be valid within nextcloud and vice versa. 

Again, I'm sorry you both had issues with upgrades, but I do plan to backport any fix we make to rc.16 so that you can then successfully upgrade.

**Comment by Firstyear** [comment]
I'm also aware that reverting the change and allowing the consent scope back is the fastest fix, and I'm tempted to take it so that we have more time for fixing it properly longer term. Currently that's where my thinking is going.

**Comment by tumbl3w33d** [comment]
Number 2 also seems fine if any benefit of the original change remains.

What you described around #2545 sounds like it would solve my [earlier confusion](https://github.com/kanidm/kanidm/issues/2596#issuecomment-1969210503) and how I imagined the client credential thing to work. So if that works for my use case, it's certainly preferable because I don't depend on a static API token mechanism in my applications. Or would this still require explicit support in the apps?

**Comment by tumbl3w33d** [comment]
There's `kanidm service-account into-person`, however, I'm not sure it works for all purposes as it's actually deprecated.

**Comment by Firstyear** [comment]
I think on rc.15 service-account into-person should still work. If not sadly it may be delete/recreate the account time? Otherwise I can help with a manual conversion script.

**Comment by Firstyear** [comment]
> Number 2 also seems fine if any benefit of the original change remains.
> 
> What you described around #2545 sounds like it would solve my [earlier confusion](https://github.com/kanidm/kanidm/issues/2596#issuecomment-1969210503) and how I imagined the client credential thing to work. So if that works for my use case, it's certainly preferable because I don't depend on a static API token mechanism in my applications. Or would this still require explicit support in the apps?

I need to email you directly about the *exact* details here, because so far you've been a little vague about some of the actors and moving parts here and without knowing precise details, it's hard to give a firm yes or no here.

**Comment by Firstyear** [comment]
After discussing in private with @tumbl3w33d I believe that client credentials grant will solve their issue once we have key domains in place. Since this won't occur immediately, I will revert this change per option 3 so that existing use cases like @tumbl3w33d's still function until we have a replacement available, and then they can smoothly transition.

**Comment by Flakebi** [comment]
> I think on rc.15 service-account into-person should still work. If not sadly it may be delete/recreate the account time? Otherwise I can help with a manual conversion script.

Unfortunately, I only get `403 Forbidden AccessDenied` when I try `into-person`.
The error is
```
requested_rem: {"api_token_session", "class", "jws_es256_private_key"} !⊆ allowed: {"account_expire", "account_valid_from", "api_token_session", "class", "devicekeys", "displayname", "gidnumber", "id_verification_eckey", "legalname", "loginshell", "mail", "name", "passkeys", "primary_credential", "radius_secret", "ssh_publickey", "unix_password", "user_auth_token_session"} | event_tag_id: 11
```

I just noticed that all accounts on my server are service accounts (and should be persons). If you have a way to convert them, that would be nice.
(It’s not that many, so one by one would be good enough.)

**Comment by Firstyear** [comment]
> > I think on rc.15 service-account into-person should still work. If not sadly it may be delete/recreate the account time? Otherwise I can help with a manual conversion script.
> 
> Unfortunately, I only get `403 Forbidden AccessDenied` when I try `into-person`. The error is
> 
> ```
> requested_rem: {"api_token_session", "class", "jws_es256_private_key"} !⊆ allowed: {"account_expire", "account_valid_from", "api_token_session", "class", "devicekeys", "displayname", "gidnumber", "id_verification_eckey", "legalname", "loginshell", "mail", "name", "passkeys", "primary_credential", "radius_secret", "ssh_publickey", "unix_password", "user_auth_token_session"} | event_tag_id: 11
> ```
> 
> I just noticed that all accounts on my server are service accounts (and should be persons). If you have a way to convert them, that would be nice. (It’s not that many, so one by one would be good enough.)

I think you may need to open a separate issue for this so I can track it and help separately

**Comment by Flakebi** [comment]
For the record, I solved this for myself by recreating my database and all groups and users.
Then I created a backup and replaced the uuids and passwords in the json with the ones from my old database.

“Restoring” from this modified backup seems to have worked, so I have no more service accounts now.

**Comment by Firstyear** [comment]
@Flakebi Really sorry for all the issues. :( We've improved a lot of how we do migrations now as a result of this incident.
